changelogs/unreleased/gh-4264-trigger-crash-on-nontrivial-change.md

```diff
@@ -0,0 +1,4 @@
+## bugfix/core
```

Comment:
There is another idea, which seem to require no additional allocations and will work both for dynamic and for static triggers. The idea is that `trigger_run()` registers self in the currently running trigger with a struct `run_link` almost like yours. The struct contains a `next_trigger` pointer. When the currently running trigger is deleted, it walks its `run_list` and updates `next_trigger` in each `run_link` to `rlist_next(self)`.
```C
struct run_link {
    struct rlist in_trigger;
    struct rlist* next_trigger;
};

void
trigger_run(struct rlist *list, void *event)
{
    struct run_link link;
    struct rlist* pos = rlist_first(list);
    while (pos != list) {
        struct trigger *t = rlist_entry(pos);
        link.next_trigger = rlist_next(t);
        rlist_append(t->run_links, &link, in_trigger);
        trigger->run(trigger, event);
        // If the trigger was deleted during `run`, then it updated `next_trigger` too.
        rlist_del(&link.in_trigger);
        pos = link.next_trigger;
    }
}
```
This is pseudocode, but I think it is quite close to how the final code could look. Looks much simpler unless I missed anything?

The behaviour difference would be is that if new triggers are added during executing the current ones, then the new ones will be executed too. If we don't want that, then we could introduce
```C
struct trigger_list {
    struct rlist triggers;
    int64_t generation;
};
```
The generation would be updated on each change of `triggers` member. `trigger_run()` would save current generation before starting to execute the triggers and skip all triggers having a newer generation. Also should be very easy to implement (except that all bare `struct rlist` trigger lists would have to be replaced by `struct trigger_list` first).

Reply:
If I understood your proposal correctly, you say that it's enough to store one `run_link` per `trigger_run`, and this `run_link` will always store the next-to-be-run trigger.

When a next to be run trigger is deleted, it updates the run_link.next_trigger to its own next trigger. And run_link starts to be "owned" by the next trigger. The trigger should point at the run link, because if it gets deleted before being run, run_link should move to the next trigger once again, and so on.

Each trigger has a list of run_links, because it might be executed recursively.

Yes, this will work without additional link allocations, this is true, and it looks nice, but I doubt it's simpler than the current approach. 

As far as I can see this will work for all cases except one: when trigger list head is swapped. Then we will loop indefinitely, because trigger->next will never be equal to old list head. This problem is present with old triggers as well, I wrote about it in the commit message:
>    3) swapping trigger list head before the last trigger is run causes an
       infinite loop or a crash (see space_swap_triggers() in alter.cc, which
       had worked all this time by miracle: space _space on_replace trigger
       swaps its own head during local recovery, and that had only worked
       because the trigger by luck was the last to run)

Also, I just noticed another problem: when the next-to-be-run trigger is last, and it is removed, run_link.next starts pointing at rlist_head. But it should distinguish the head from the other elements somehow. Otherwise it would try to make rlist_head "own" the run_link. So the run_link has to remember list head, and this won't work with swapping trigger list head once again.

So it isn't that easy, when we go into details, and I'd personally stay with the current approach.

Reply:
It would work simple though, if you would make `run_link` in this approach store not just `next_trigger` as direct `rlist_next()` but make it `NULL` if the deleted trigger realized that it is the last one. I suppose we can return to this if any benchmarks would in future reveal a slowdown because of the hot path allocations.